CMakeLists.txt

```diff
@@ -71,6 +71,7 @@ option(BUILD_CUSTOM_PROTOBUF "Build and use Caffe2's own protobuf under third_pa
 option(BUILD_PYTHON "Build Python binaries" ON)
 option(BUILD_CAFFE2_OPS "Build Caffe2 operators" ON)
 option(BUILD_SHARED_LIBS "Build libcaffe2.so" ON)
+option(BUILD_TORCH_MOBILE "Build libtorch for mobile" OFF)
```

Comment:
There's BUILD_ATEN_MOBILE flag above - is it relevant to expose it to end user?

Reply:
Yup - can we unify them? Also, seems like `BUILD_MOBILE` might be the best name if you're willing to tweak it.

Reply:
Please add this to cmake/Summary.cmake as well so it's shown in the CMake output.

Reply:
BUILD_ATEN_MOBILE was not introduced by me so I can only guess its meaning by reading the code - from what I see looks like it's always set when platform is Android/iOS, so it means "we are building mobile library".

Currently with BUILD_ATEN_MOBILE=True it only builds ATen/Core but not builds actual ATen kernels nor TH/THNN kernels.

What this PR does it to include ATen & TH kernels when a new macro is set to true. So there are a few design decisions we need think about here:

1. Do we want to expose the switch of building ATen/TH kernels to end user?
I guess the answer is yes? - if current caffe2 mobile library is still being used by our users. Always building ATen/TH kernels will regress caffe2 mobile library build size until we provide a solution to modularize ATen/TH kernels.

2. How do we structure the build macros?

A - doesn't change BUILD_ATEN_MOBILE meaning, introducing BUILD_TORCH_MOBILE to enable ATen/TH kernels (what this PR does):
```
+-----------------+-------------------+--------------------+
|                 | BUILD_ATEN_MOBILE | BUILD_TORCH_MOBILE |
+-----------------+-------------------+--------------------+
| Mobile platform | ON                |                    |
| Build caffe2go  | ON                | OFF                |
| Build libtorch  | ON                | ON                 |
+-----------------+-------------------+--------------------+
```

B - replace BUILD_ATEN_MOBILE with BUILD_MOBILE, which still enabled for all mobile builds; use BUILD_ATEN (instead of BUILD_TORCH_MOBILE) to build ATen/TH kernels:
```
+-----------------+--------------+------------+-------------+
|                 | BUILD_MOBILE | BUILD_ATEN | BUILD_TORCH |
+-----------------+--------------+------------+-------------+
| Mobile platform | ON           |            |             |
| Build caffe2go  | ON           | OFF        | OFF         |
| Build libtorch  | ON           | ON         | ON          |
+-----------------+--------------+------------+-------------+
```
Note that BUILD_TORCH is an existing build macro - we don't need change it for now. Maybe add BUILD_JIT or something else later to provide fine-grain control for libtorch mobile build.

How does 2B sounds to you?
@kostmo @dzhulgakov @orionr

Reply:
2B sounds reasonable to me! Thanks @ljk53.

Reply:
Updated the PR according to 2B with a few modifications:
- rename BUILD_ATEN_MOBILE to INTERN_BUILD_MOBILE and hide it from users;
- use INTERN_BUILD_ATEN_OPS to control whether build aten ops and set it according to INTERN_BUILD_MOBILE & BUILD_TORCH_MOBILE;
- keep BUILD_TORCH_MOBILE to differentiates caffe2 mobile build and libtorch mobile build;
- How we are going to use BUILD_TORCH_MOBILE:
1. as a master switch BUILD_TORCH_MOBILE automatically enables/disables other macros, e.g.: BUILD_TORCH / BUILD_CAFFE2_OPS;
2. strip out some ATen/TH/Caffe2 code that are not applicable for mobile - directly using INTERN_BUILD_MOBILE for this purpose will affect caffe2 mobile build, e.g. removing threadpool on this PR;

Reply:
Eventually, I think Caffe2 mobile builds shouldn't exist. Given that, I'm still tempted to call it BUILD_MOBILE. We can chat offline, though.

Reply:
Chatted with @orionr offline - we decided to flip and rename BUILD_TORCH_MOBILE to BUILD_CAFFE2_MOBILE and mark it as deprecating. Eventually libtorch will be default mobile build. Keep the BUILD_CAFFE2_MOBILE flag just to preserve build_android.sh functionality for now.